GameLogic/python_emulator.py

```diff
@@ -0,0 +1,531 @@
+            )
+        half_elapsed_time = elapsed_time - self.time_to_peak
+        return self.JUMP_HEIGHT - int(0.5 * self.acceleration * half_elapsed_time**2)
+
+    # If possible, an item (cactus, mid bird) will be created and the cooldown will start
+    def spawn_item(self):
+        if self.cannot_spawn_object:
+            return
+        # Otherwise we can spawn a cactus right now
+        random_chance = random.uniform(0, 1)
+        if random_chance < self.CACTUS_CHANCE:
+            self.cactus_locs.append(self.WIDTH)
+            self.cannot_spawn_object = True
+            self.cooldown_start_time = pygame.time.get_ticks()
+        elif random_chance < self.MID_BIRD_CHANCE:
+            self.mid_bird_locs.append(self.WIDTH)
+            self.cannot_spawn_object = True
+            self.cooldown_start_time = pygame.time.get_ticks()
+        else:
+            return
+        return
+
+    #############################################################################
+    # Main Loop
+    #############################################################################
+
+    def run(self):
+
+        # Wait until first space bar pess until the game starts
+        while True:
+            for event in pygame.event.get():
+                if event.type == pygame.KEYDOWN and event.key == pygame.K_SPACE:
+                    self.start_game = True
+                    self.jump_start_time = pygame.time.get_ticks()
+            if self.start_game:
+                break
+            self.screen.fill(self.BG_COLOR)
+            self.draw_ground()
+            self.determine_and_draw_dino()
+            pygame.display.flip()
+            self.clock.tick(self.FPS)
+
+        # Simulates the whole loop running
+        while True:
+
+            # First we detect to see if anything was triggered
+            for event in pygame.event.get():
+                # Closes the system window
+                if event.type == pygame.QUIT:
+                    pygame.quit()
+                    sys.exit()
+                # Checks for key press events
+                if event.type == pygame.KEYDOWN:
+                    # Dino Jump is triggered (space bar)
+                    if (
+                        event.key == pygame.K_SPACE
+                        and not self.is_jumping
+                        and not self.is_ducking
+                    ):
+                        self.is_jumping = True
+                        self.jump_start_time = pygame.time.get_ticks()
+                    # Dino duck is triggered (down key)
+                    if event.key == pygame.K_DOWN and not self.is_jumping:
+                        self.is_ducking = True
+                # Checks for key release events
+                if event.type == pygame.KEYUP:
+                    # Dino duck is released (down key)
+                    if event.key == pygame.K_DOWN:
+                        self.is_ducking = False
+
```

Comment:
I think the code here is pretty readable and so it might actually read better without the comments.